stellargraph/datasets/dataset_loader.py

```diff
@@ -0,0 +1,195 @@
+        NOTE: this is not compatible with python's ABC abstract base class, so this class derives from object."""
+        cls.name = name
+        cls.directory_name = directory_name
+        cls.url = url
+        cls.url_archive_format = url_archive_format
+        cls.expected_files = expected_files
+        cls.description = description
+        cls.source = source
+        cls.data_subdirectory_name = data_subdirectory_name
+
+        if url_archive_format is None and len(expected_files) != 1:
+            raise ValueError(
+                "url_archive_format is None, which requires a single expected_file, found: {expected_files!r}"
+            )
+
+        # auto generate documentation
+        if cls.__doc__ is not None:
+            raise ValueError(
+                "DatasetLoader docs are automatically generated and should be empty"
+            )
+        cls.__doc__ = f"{cls.description}\n\nFurther details at: {cls.source}"
+
+        super().__init_subclass__(**kwargs)  # type: ignore
+
+    def __init__(self) -> None:
+        # basic check since this is effectively an abstract base class, and derived classes should have set name
+        if not self.name:
+            raise ValueError(
+                f"{self.__class__.__name__} can't be instantiated directly, please use a derived class"
+            )
+
+    @property
+    def base_directory(self) -> str:
+        """str: The full path of the directory containing this dataset."""
+        return os.path.join(self._all_datasets_directory(), self.directory_name)
+
+    @property
+    def data_directory(self) -> str:
+        """str: The full path of the directory containing the data content files for this dataset."""
+        if self.data_subdirectory_name is None:
+            return self.base_directory
+        else:
+            return os.path.join(self.base_directory, self.data_subdirectory_name)
+
+    def _create_base_directory(self) -> None:
+        data_dir = self.base_directory
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+    @staticmethod
+    def _all_datasets_directory() -> str:
+        """Return the path of the base directory which contains subdirectories for each dataset."""
+        return os.path.expanduser(
+            os.getenv(
+                "STELLARGRAPH_DATASETS_PATH", os.path.join("~", "stellargraph-datasets")
+            )
+        )
+
+    def _resolve_path(self, filename: str) -> str:
+        """Convert dataset relative file names to their full path on filesystem"""
+        return os.path.join(self.base_directory, filename)
+
+    def _is_downloaded(self) -> bool:
+        """Returns true if the expected files for the dataset are present"""
+        return all(
+            os.path.isfile(self._resolve_path(file)) for file in self.expected_files
+        )
+
+    def _verify_files_downloaded(self) -> None:
```

Comment:
This and `is_downloaded` have practically the same core. Maybe:

```python
def _missing_files(self) -> List[str]:
    return [file for file in self.expected_files if not ....]

def _is_downloaded(self) -> bool:
    return len(self._missing_files()) == 0

def _verify_files_downloaded(self) -> None:
    missing = self._missing_files()
    if missing:
        raise FileNotFoundError(...)
```

With the `_missing_files()` method, we potentially don't need `_verify_files_downloaded` because it's small enough to just inline into `download`.